### ISSUE TITLE
Review new functionality specific to C* 4.0.0 for impact within k8ssandra

### ISSUE THREAD

**Issue by jdonenine** [k8ssandra/k8ssandra#484]
**Is your feature request related to a problem? Please describe.**
N/A

**Describe the solution you'd like**
Review the new functionality coming available in C* 4.0.0 to identify any impact that it might have on k8ssandra and identify areas where we may need to adjust our deployment or expose new configuration options to control new and useful features.

The focus here isn't to make changes, but to outline proposed enhancements or required changes in k8ssandra issues to be included in planning as 4.0.0 approaches GA.

**Comment by adejanovski** [comment]
Here's a GH discussion with the list of 4.0 features I think are of interest: https://github.com/k8ssandra/k8ssandra/discussions/565

**Comment by jdonenine** [comment]
We've shut down the discussion in GH in favor of the forums, so I'm bringing some of the content from the discussion mentioned in the comments above here.

Here is the original content from @adejanovski 

## Enhanced token allocation algorithm
16 vnodes by default.
`allocate_tokens_for_replication_factor` yaml setting, in replacement of the much harder to use `allocate_tokens_for_keyspace` setting.

## Incremental repair fix
Check [my blog post](https://thelastpickle.com/blog/2018/09/10/incremental-repair-improvements-in-cassandra-4.html) for more information on the improvements.
Add Reaper auto-scheduling based on % repaired, using incremental repair.
This could allow to improve tombstone purges safely by:

- Automatically and quickly repairing small chunks of data
- Allowing the purge of tombstones that have been repaired only (cassandra.yaml setting)
- Which allows to reduce gc_grace_seconds to the hint window without fearing zombie data

We should also allow enabling aggressive purging of fully expired sstables in TWCS (requires a specific JVM flag and to allow enabling through a compaction subproperty on individual tables).



## Compaction improvements
Higher compaction throughput by default. Maybe we don’t need to care about this unless we don’t use the new defaults seamlessly.

## JDK11 support with Shenandoah & ZGC
[Benchmarks](https://thelastpickle.com/blog/2020/06/29/cassandra_4-0_garbage_collectors_performance_benchmarks.html) comparing 3.11 vs 4.0~alpha4 showed tremendous improvement in throughput and especially latencies (-88% at p99 when using Shenandoah).

Shenandoah is available in JDK11 using specific builds:

```
Releases
Shenandoah availability differs by vendor and JDK release. OpenJDK 12+ builds normally include Shenandoah by default. OpenJDK 11 requires the opt-in during build time.

Known vendor status is:

Red Hat
Fedora 24+ OpenJDK 8+ builds include Shenandoah

RHEL 7.4+ ships with OpenJDK 8+ that includes Shenandoah as Technology Preview
Red Hat OpenJDK 8u builds for Windows include Shenandoah
Amazon
Ships Shenandoah in Amazon Corretto, starting with OpenJDK 11.0.9
Oracle
Does not ship Shenandoah in any release, both OpenJDK builds and proprietary builds
Azul
Ships Shenandoah in Azul Zulu, starting with OpenJDK 11.0.9
SAP
(status unknown)
AdoptOpenJDK
Ships Shenandoah in default binaries, starting with OpenJDK 11.0.9
Linux distributions
Debian ships Shenandoah starting with OpenJDK 11.0.9
Gentoo ebuilds for IcedTea have Shenandoah USE flag
RHEL/Fedora-based distros or other distros that use packages from them may also have Shenandoah enabled. Notably, CentOS, Oracle Linux, and Amazon Linux are known to ship it.
```

I assume this would require us to create custom Apache Cassandra docker images to use with alternate JDKs.
Performance wise this is something that should be seriously considered though.

## Client Backpressure
To avoid having nodes crashing to OOM, client backpressure was implemented in 4.0.
[This blog post](https://cassandra.apache.org/blog/2020/09/03/improving-resiliency.html) describes the issue and how 4.0 solves it. 

On the server side, this feature exposes two new thresholds in cassandra.yaml:

```
native_transport_max_concurrent_requests_in_bytes_per_ip
native_transport_max_concurrent_requests_in_bytes
```

Not sure for now how these should be configured, but exposing them would be a good thing.


## Audit + Full Query Logging
Audit logging is a nice feature to comply with some regulations such as SOX, and log every interaction between clients and clusters (connection attempts, failed/successful queries, etc…).

Full query logging allows to dump all queries going through a node to a binary file, which can be used for replay or analysis.
